File: backend/src/protocol.rs
use serde::{Deserialize, Serialize};

const TYPE_DATA: u8 = 0x00;
const TYPE_RESIZE: u8 = 0x01;
const TYPE_PING: u8 = 0x02;
const TYPE_PONG: u8 = 0x03;
const TYPE_SESSION_EVENT: u8 = 0x04;

#[derive(Debug, PartialEq)]
pub enum Frame {
    Data(Vec<u8>),
    Resize { cols: u16, rows: u16 },
    Ping,
    Pong,
    SessionEvent(SessionEventPayload),
}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub struct SessionEventPayload {
    pub event_type: String,
    pub session_id: String,
}

impl Frame {
    pub fn encode(&self) -> Vec<u8> {
        match self {
            Frame::Data(data) => {
                let mut buf = Vec::with_capacity(1 + data.len());
                buf.push(TYPE_DATA);
                buf.extend_from_slice(data);
                buf
            }
            Frame::Resize { cols, rows } => {
                let mut buf = Vec::with_capacity(5);
                buf.push(TYPE_RESIZE);
                buf.extend_from_slice(&cols.to_be_bytes());
                buf.extend_from_slice(&rows.to_be_bytes());
                buf
            }
            Frame::Ping => vec![TYPE_PING],
            Frame::Pong => vec![TYPE_PONG],
            Frame::SessionEvent(payload) => {
                let json = serde_json::to_vec(payload).unwrap();
                let mut buf = Vec::with_capacity(1 + json.len());
                buf.push(TYPE_SESSION_EVENT);
                buf.extend_from_slice(&json);
                buf
            }
        }
    }

    pub fn decode(data: &[u8]) -> Result<Self, String> {
        if data.is_empty() {
            return Err("Empty frame".to_string());
        }
        match data[0] {
            TYPE_DATA => Ok(Frame::Data(data[1..].to_vec())),
            TYPE_RESIZE => {
                if data.len() < 5 {
                    return Err("Resize frame too short".to_string());
                }
                let cols = u16::from_be_bytes([data[1], data[2]]);
                let rows = u16::from_be_bytes([data[3], data[4]]);
                Ok(Frame::Resize { cols, rows })
            }
            TYPE_PING => Ok(Frame::Ping),
            TYPE_PONG => Ok(Frame::Pong),
            TYPE_SESSION_EVENT => {
                let payload: SessionEventPayload = serde_json::from_slice(&data[1..])
                    .map_err(|e| format!("Invalid session event: {}", e))?;
                Ok(Frame::SessionEvent(payload))
            }
            t => Err(format!("Unknown frame type: 0x{:02x}", t)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn data_frame_roundtrip() {
        let frame = Frame::Data(b"hello terminal".to_vec());
        let encoded = frame.encode();
        assert_eq!(encoded[0], TYPE_DATA);
        let decoded = Frame::decode(&encoded).unwrap();
        assert_eq!(decoded, frame);
    }

    #[test]
    fn resize_frame_roundtrip() {
        let frame = Frame::Resize { cols: 120, rows: 40 };
        let encoded = frame.encode();
        assert_eq!(encoded[0], TYPE_RESIZE);
        assert_eq!(encoded.len(), 5);
        let decoded = Frame::decode(&encoded).unwrap();
        assert_eq!(decoded, frame);
    }

    #[test]
    fn ping_pong_roundtrip() {
        let ping = Frame::Ping;
        assert_eq!(Frame::decode(&ping.encode()).unwrap(), Frame::Ping);
        let pong = Frame::Pong;
        assert_eq!(Frame::decode(&pong.encode()).unwrap(), Frame::Pong);
    }

    #[test]
    fn session_event_roundtrip() {
        let frame = Frame::SessionEvent(SessionEventPayload {
            event_type: "ended".to_string(),
            session_id: "rc-abc123".to_string(),
        });
        let encoded = frame.encode();
        assert_eq!(encoded[0], TYPE_SESSION_EVENT);
        let decoded = Frame::decode(&encoded).unwrap();
        assert_eq!(decoded, frame);
    }

    #[test]
    fn empty_frame_returns_error() {
        assert!(Frame::decode(&[]).is_err());
    }

    #[test]
    fn unknown_type_returns_error() {
        assert!(Frame::decode(&[0xFF]).is_err());
    }

    #[test]
    fn resize_too_short_returns_error() {
        assert!(Frame::decode(&[TYPE_RESIZE, 0x00]).is_err());
    }
}

File: android/app/src/main/kotlin/com/remotecontrol/data/websocket/Protocol.kt
package com.remotecontrol.data.websocket

import com.google.gson.Gson
import java.nio.ByteBuffer

sealed class Frame {
    data class Data(val payload: ByteArray) : Frame() {
        override fun equals(other: Any?) = other is Data && payload.contentEquals(other.payload)
        override fun hashCode() = payload.contentHashCode()
    }
    data class Resize(val cols: Int, val rows: Int) : Frame()
    data object Ping : Frame()
    data object Pong : Frame()
    data class SessionEvent(val eventType: String, val sessionId: String) : Frame()
    data class Auth(val token: ByteArray) : Frame() {
        override fun equals(other: Any?) = other is Auth && token.contentEquals(other.token)
        override fun hashCode() = token.contentHashCode()
    }

    fun encode(): ByteArray = when (this) {
        is Data -> byteArrayOf(TYPE_DATA) + payload
        is Resize -> {
            val buf = ByteBuffer.allocate(5)
            buf.put(TYPE_RESIZE)
            buf.putShort(cols.toShort())
            buf.putShort(rows.toShort())
            buf.array()
        }
        is Ping -> byteArrayOf(TYPE_PING)
        is Pong -> byteArrayOf(TYPE_PONG)
        is SessionEvent -> {
            val json = gson.toJson(mapOf("event_type" to eventType, "session_id" to sessionId))
            byteArrayOf(TYPE_SESSION_EVENT) + json.toByteArray()
        }
        is Auth -> byteArrayOf(TYPE_AUTH) + token
    }

    companion object {
        private const val TYPE_DATA: Byte = 0x00
        private const val TYPE_RESIZE: Byte = 0x01
        private const val TYPE_PING: Byte = 0x02
        private const val TYPE_PONG: Byte = 0x03
        private const val TYPE_SESSION_EVENT: Byte = 0x04
        private const val TYPE_AUTH: Byte = 0x05
        private val gson = Gson()

        fun decode(data: ByteArray): Frame {
            require(data.isNotEmpty()) { "Empty frame" }
            return when (data[0]) {
                TYPE_DATA -> Data(data.copyOfRange(1, data.size))
                TYPE_RESIZE -> {
                    require(data.size >= 5) { "Resize frame too short" }
                    val buf = ByteBuffer.wrap(data, 1, 4)
                    Resize(cols = buf.short.toInt() and 0xFFFF, rows = buf.short.toInt() and 0xFFFF)
                }
                TYPE_PING -> Ping
                TYPE_PONG -> Pong
                TYPE_SESSION_EVENT -> {
                    val json = String(data, 1, data.size - 1)
                    @Suppress("UNCHECKED_CAST")
                    val map = gson.fromJson(json, Map::class.java) as Map<String, String>
                    SessionEvent(eventType = map["event_type"] ?: "", sessionId = map["session_id"] ?: "")
                }
                TYPE_AUTH -> Auth(data.copyOfRange(1, data.size))
                else -> throw IllegalArgumentException("Unknown frame type: 0x${String.format("%02x", data[0])}")
            }
        }
    }
}
